common-npm-packages/docker-common-v2/registryauthenticationprovider/acrauthenticationtokenprovider.ts

```
@@ -45,4 +48,56 @@ export default class ACRAuthenticationTokenProvider extends AuthenticationTokenP
+                if (response.statusCode === 200) {
+                    deferred.resolve(response.body.refresh_token);
+                }
+                else if (response.statusCode == 429 || response.statusCode == 500) {
+                    if (retryCount < retryLimit) {
+                        let waitedTime = 2000 + timeToWait * 2;
+                        retryCount += 1;
+                        setTimeout(() => {
+                            deferred.resolve(this._getACRToken(AADToken, retryCount, waitedTime));
+                        }, waitedTime);
+                    }
+                    else {
+                        deferred.reject(tl.loc('CouldNotFetchAccessTokenforACRStatusCode', response.statusCode, response.statusMessage));
+                    }
+
+                }
+                else {
+                    deferred.reject(tl.loc('CouldNotFetchAccessTokenforMSIDueToACRNotConfiguredProperlyStatusCode', response.statusCode, response.statusMessage));
+                }
+            },
+            (error) => {
+                deferred.reject(error)
+            }
+        );
+        return deferred.promise;
+    }
+
+    public async getMSIAuthenticationToken(retryCount: number, timeToWait: number): Promise<RegistryAuthenticationToken> {
+        if (this.registryURL && this.endpointName) {
+            let azureResourceManagerCommon1 = new azureResourceManagerCommon.ApplicationTokenCredentials("", "tempDomain", "", "https://management.core.windows.net/", "tempAuthorityURL", "tempActiveDirectoryResourceId", false, "ManagedServiceIdentity", null, null, null, null, null)
+            let aadtoken = await azureResourceManagerCommon1.getToken();
+            let acrToken = await this._getACRToken(aadtoken, retryCount, timeToWait);
+            return new RegistryAuthenticationToken("00000000-0000-0000-0000-000000000000", acrToken, this.registryURL, "ManagedIdentity@AzureRM", this.getXMetaSourceClient());
+        }
+        return null;
```

When this code can get hit ? I would prefer to throw rather than causing null ref in caller

---

I tried to replicate the SPN login here, in that, if there is not enough information that is provided, the code will not work/run and return a null pointer, or in the updated case, throw an error.